content/articles/php-io-streams/index.md

```diff
@@ -0,0 +1,191 @@
+Streams have implementation wrappers that allow you to handle specific protocols.
+
+### The file system handler
+To understand the file system handler, let's create a file called `test.txt` in the `/tmp` directory.
+
+Proceed and add texts in the file, in our case, `test.txt` will contain the following text:
+```text
+hello world
+```
+
+Next, open `streams.php` file and add the following code:
+```php
+<?php
+$content = file_get_contents(__DIR__ . '/test.txt');
+echo $content;
+```
+
+In the above code, we are using the `file_get_contents()` function to read the file's contents.
+
+When we run the above code, we get the following result:
+```bash
+hello world
+```
+
+You notice that the code runs smoothly. However, this will not always be the case while handling large files as shown below:
+
+```php
+<?php
+$content = file_get_contents(__DIR__ . '/large-test-file.txt');
+echo $content;
+```
+
+When the above code is run, you will get the following error:
+```bash
+# php fatal error:  Allowed memory size of 33554432 bytes exhausted (tried to allocate 524288 bytes) in /tmp/streams.php on line 3
+PHP Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 1048576000 bytes)
+```
+
+The above error is because the file is too large to be loaded into memory. As we may recall, PHP has the memory limit set to `128M`.
+
+Now that we understand the problem of handling large files, let's introduce the concept of streams in the next section.
+
+### Working with streams
+Previously, we have seen how PHP default configurations restrict us to work within a set memory limit.
+
+In this section, let's first create a sample script to solve our above error using the concept of streams.
+```php
+//open the stream.php file and add the following:
+<?php
+$large_file = fopen(__DIR__ . '/large-test-file.txt', 'r');
+//the result is printed until the end of the file
+while(! feof($large_file)) {
+  $endLine = fgets($large_file);
+  echo $endLine. "<br>";
+}
+fclose($large_file);
+```
+
+In the above code, we use the `fopen()` function to open the file. The `fopen()` function takes in 2 parameters. The first is the file path, and the second is the mode.
+
+Next, we are using the `fgets()` function to read the file's contents.
+Finally, we use the `fclose()` function to close the file.
+
+When you execute the above code, you notice that it doesn't give you an error, as was the case when we used the `file_get_contents()` function. This is the power of streams.
+
+With streams, we are able to read any file size without running into memory management problems which maybe very frustrating to a developer.
+
+### Retrieving file contents
+In the previous section, we have seen how we can use the power of streams in PHP to handle large files, in our case, using the `fopen()` function. 
+
+Now, let's understand how we can retrieve the file's contents using the `fgets()` function.
+```php
+...
```

Comment:
check this code section. do we need the extra ... before and after the code snippets?

Reply:
yes, we need them, to show that some imports are needed for the code to work!